mtl/psmtl2milp.py

```diff
@@ -0,0 +1,172 @@
+            
+        if t not in self.variables[state]:
+            self.variables[state][t] = z
+            self.model.update()
+        return self.variables[state][t]
+
+    def predicate(self, pred, z, t, depth, z_ancestors):
+        '''Adds a predicate to the model.'''
+        assert pred.op == Operation.PRED
+        self.add_state(pred.variable, t, z)
+        
+
+    def conjunction(self, formula, z, t, depth, z_ancestors):
+        '''Adds a conjunction to the model.'''
+        assert formula.op == Operation.AND
+        z_children = [self.to_milp(f, t, depth+1, z_ancestors) 
+                      for f in formula.children]
+        self.model.addConstr(z == sum(z_children) / len(z_children) )
+
+    def disjunction(self, formula, z, t, depth, z_ancestors):
+        '''Adds a disjunction to the model.'''
+        assert formula.op == Operation.OR
+        z_children = [self.to_milp(f, t, depth+1, z_ancestors) 
+                      for f in formula.children]
+        self.model.addConstr(z == grb.max_(z_children))
+
+    def eventually(self, formula, z, t, depth, z_ancestors):
+        '''Adds an eventually to the model.'''
+        assert formula.op == Operation.EVENT
+        a, b = int(formula.low), int(formula.high)
+        child = formula.child
+        z_children = [self.to_milp(child, t + tau, depth+1, z_ancestors) 
+                      for tau in range(a, b+1)]
+        self.model.addConstr(z == grb.max_(z_children))
+
+    def globally(self, formula, z, t, depth, z_ancestors):
+        '''Adds a globally to the model.'''
+        assert formula.op == Operation.ALWAYS
+        a, b = int(formula.low), int(formula.high)
+        child = formula.child
+        z_children = [self.to_milp(child, t + tau, depth+1, z_ancestors) 
+                      for tau in range(a, b+1)]
+        self.model.addConstr(z == sum(z_children) / (b-a+1))    
+
+    def until(self, formula, z, t, depth, z_ancestors):
+        '''Adds an until to the model.'''
+        a, b = int(formula.low), int(formula.high)
+        z_children = []
+        for t_ in range(a,b+1):
+            z_children_left =  [self.to_milp(formula.left, t+t__, depth+1, 
+                                             z_ancestors) 
+                                             for t__ in range(0,t_)]
+            z_children_right = [self.to_milp(formula.right, t+t_, depth+1, 
+                                             z_ancestors)]
+            z_children.append(z_children_right + sum(z_children_left))
+
+        self.model.addConstr(z == grb.max_(z_children))
+
+    def hierarchical(self): #(Hierarchical Optimization)
+        max_depth = max(self.objectives)
+        for d in range(max_depth+1):
+            self.model.setObjectiveN(-self.objectives[d], d, 
+                                     priority=max_depth-d)
+            self.model.update()
+        
+        self.model.optimize()
+        self.model.write('model_test.lp')
```

Comment:
Might be better to have the model output filename a parameter of the method with the default value of `'model_test.lp'`.